### PR TITLE
/go/libraries/doltcore/sqle/writer: Extracted Interface for WriteSession

### PR DESCRIPTION
**Pull request by andrew-wm-arthur**
Extracted an interface from TableEditSession to allow for multiple implementations for different formats

### CODE REVIEW

File: go/libraries/doltcore/sqle/writer/table_editor.go [hunk 1]
```diff
@@ -41,17 +40,19 @@ type TableEditor interface {
 	NextAutoIncrementValue(potentialVal, tableVal interface{}) (interface{}, error)
 }
 
-// sqlTableEditor is a wrapper for *doltdb.SessionedTableEditor that complies with the SQL interface.
+type SessionRootSetter func(ctx *sql.Context, dbName string, root *doltdb.RootValue) error
```
Comment (zachmu) [inline go/libraries/doltcore/sqle/writer/table_editor.go:43]: Probably should go higher in the file, with a type comment

Comment (zachmu) [reply go/libraries/doltcore/sqle/writer/table_editor.go:43]: These days I'm wary of defining top-level callback types like this, and especially embedding them in other structs like you do below. In general it seems like a sign that we're missing some other load-bearing abstraction. My rule of thumb is that function entities belong as params to methods only, almost never as fields.

File: go/libraries/doltcore/table/editor/bulk_import_tea.go [hunk 2]
```diff
@@ -51,6 +51,19 @@ func (cap *fixedCapacity) capacityExceeded(size int) bool {
 
 var _ TableEditAccumulator = (*BulkImportTEA)(nil)
 
+// Options are properties that define different functionality for the tableEditSession.
+type Options struct {
```
Comment (zachmu) [inline go/libraries/doltcore/table/editor/bulk_import_tea.go:55]: This seems misplaced, is there a better file for this to live in?

**Review by zachmu — APPROVED** [review]
Generally like the direction this moves us in, just a few comments

File: go/libraries/doltcore/sqle/writer/table_editor.go (hunk 1 above)
Comment (reltuk) [reply go/libraries/doltcore/sqle/writer/table_editor.go:43]: In general I think I agree that that often naked function types can point to some missing abstraction that would actual help structure things better. But I did want to point out that embedding function fields in structs to provide pluggable functionality is common throughout the go standard library and I think it's a legitimate and reasonable approach. See `http.Client`, `http.Transport`, `tls.Config`, etc.

Comment (andrew-wm-arthur) [reply go/libraries/doltcore/sqle/writer/table_editor.go:43]: In this case it points to a flaw in abstractions. I'm still working out what exactly is missing. 

`SessionRootSetter` is set to `dsess.Session.SetRoot()`, which currently needs to be called each time we `Close()` a sqlTableEditor.

**Pull request merged**